### PR TITLE
make jdao removeAll Sink aware

### PR DESCRIPTION
**Pull request by drish**
#484

### CODE REVIEW

**Comment by drish** [comment]
not ready yet to be merged
wip

File: src/foam/dao/RemoveSink.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package foam.dao;
+
+import foam.core.Detachable;
+import foam.core.FObject;
+import foam.core.X;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+
+public class RemoveSink {
```
Comment (kgrgreer) [inline src/foam/dao/RemoveSink.java]: This isn't a Sink since it doesn't implement the Sink interface.

File: src/foam/dao/JDAO.java [hunk 2]
```diff
@@ -50,10 +51,17 @@ public FObject remove_(X x, FObject obj) {
   }
 
   @Override
-  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
+  public void removeAll_(final X x, long skip, final long limit, Comparator order, Predicate predicate) {
     // TODO: this is wrong, should only call journal.removeAll() if neither limit nor predicate
     // are set.
-    journal.removeAll();
+
+    getDelegate().select_(x, new AbstractSink() {
```
Comment (kgrgreer) [inline src/foam/dao/JDAO.java]: Instead of doing "new AbstractSink()" you should be doing "new RemoveSink()".

File: src/foam/dao/RemoveSink.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  public RemoveSink(final DAO dao) {
```
Comment (kgrgreer) [inline src/foam/dao/RemoveSink.java]: RemoveSink's constructor should just store the 'dao' and do nothing else. It should have a put() method that does dao.remove(x, obj);

File: src/foam/dao/RemoveSink.java [hunk 4]
```diff
@@ -6,25 +6,32 @@
...
+
+  protected DAO dao;
+
+  public RemoveSink(DAO dao) {
+    this.dao = dao;
+  }
+
+  @Override
+  public void put(FObject obj, Detachable sub) {
+    final X x = ((AbstractDAO) this.dao).getX();
```
Comment (kgrgreer) [inline src/foam/dao/RemoveSink.java]: Replace these two lines with just: dao.remove(obj);

dao already has a context and you don't need "this." in Java.

File: src/foam/dao/RemoveSink.java [hunk 5]
```diff
@@ -6,25 +6,32 @@
...
 
-  public RemoveSink(final DAO dao, long skip, long limit, Comparator order, Predicate predicate) {
-    final X x = ((AbstractDAO) dao).getX();
-    dao.select_(x, new AbstractSink() {
-      @Override
-      public void put(FObject obj, Detachable sub) {
-        dao.remove_(x, obj);
-      }
-    }, skip, limit, order, predicate);
+  @Override
```
Comment (kgrgreer) [inline src/foam/dao/RemoveSink.java]: If you extend AbstractSink, then you don't need to implement these NOP methods.

Comment (drish) [reply src/foam/dao/RemoveSink.java]: oh ok, thanks. 
done.

File: src/foam/dao/JDAO.java [hunk 6]
```diff
@@ -50,10 +51,11 @@ public FObject remove_(X x, FObject obj) {
   }
 
   @Override
-  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
+  public void removeAll_(final X x, long skip, final long limit, Comparator order, Predicate predicate) {
```
Comment (kgrgreer) [inline src/foam/dao/JDAO.java:54]: If you move this to AbstractDAO instead of JDAO, then it will fix MapDAO also.

**Pull request merged**